prov/hook/profile/src/hook_profile.c

```diff
@@ -0,0 +1,906 @@
+/*
+ * Copyright (c) 2018-2023 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "ofi_hook.h"
+#include "ofi_prov.h"
+#include "ofi_iov.h"
+#include "hook_prov.h"
+
+#include "hook_profile.h"
+
+static inline struct profile_context *profile_ctx(struct hook_ep *ep)
+{
+	return &container_of(ep->domain->fabric, struct profile_fabric,
+	                     fabric_hook)->prof_ctx;
+}
+
+static inline struct profile_context *profile_ctx_cq(struct hook_cq *cq)
+{
+	return &container_of(cq->domain->fabric, struct profile_fabric,
+	                     fabric_hook)->prof_ctx;
+}
+
+static inline struct profile_context *profile_ctx_domain(struct hook_domain *dom)
+{
+	return &container_of(dom->fabric, struct profile_fabric,
+	                     fabric_hook)->prof_ctx;
+}
+
+static inline int prof_size_bucket(size_t len)
+{
+	if (len <= 0x1000) {  // 4K
+		if (len <= 0x40)
+			return PROF_SIZE_0_64;
+		if (len <= 0x200)
+			return PROF_SIZE_64_512;
+		if (len <= 0x400)
+			return PROF_SIZE_512_1K;
+		else
+			return PROF_SIZE_1K_4K;
```

Comment:
There's no reason for nesting the if statements.  The code would be easier to read using the base-10 values, rather than hex values.  E.g.
```
if (len <= 64)
    ...
if (len <= 512)
    ...
```